src/ring_theory/discriminant.lean

```diff
@@ -0,0 +1,221 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import linear_algebra.matrix.determinant
+import ring_theory.trace
+import ring_theory.norm
+import linear_algebra.vandermonde
+
+/-!
+# Discriminant.
+Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define the
+*discriminant* of `b` as the determinant of the matrix whose `(i j)`-th element is the trace of
+`b i * b j`.
+
+## Main definition
+* `discriminant A b` : the discriminant of `b : ι → B`.
+
+## Main results
+* `zero_of_not_linear_independent` : if `b` is not linear independent, then
+  `discriminant A b = 0`.
+* `of_matrix_vec_mul` and `of_matrix_mul_vec` : formulas relating `discriminant A ι b` with
+  `discriminant A ((P.map (algebra_map A B)).vec_mul b)` and
+  `discriminant A ((P.map (algebra_map A B)).mul_vec b)`.
+* `not_zero_of_linear_independent` : over a field, if b` is linear independent, then
+  `discriminant K b ≠ 0`.
+* `eq_det_embeddings_matrix_reindex_pow_two` if `L/K` is a fields extension and `b : ι → L`, then
+  `discriminant K b` is the square of the determinant of the matrix whose `(i, j)` coefficient is
+  `σⱼ (b i)`, where `σⱼ : L →ₐ[K] E` is the embedding in an algebraically closed field `E`
+  corresponding to `j : ι` via a bijection `e : ι ≃ (L →ₐ[K] E)`.
+* `of_power_basis_eq_prod` : the discriminant of a power basis.
+
+## Implementation details
+Our definition works for any `A`-algebra `B`, but note that if `B` is not free as an `A`-module,
+then `trace A B = 0` by definition, so `discriminant A b = 0` for any `b`.
+-/
+
+universes u v w z
+
+open_locale matrix big_operators
+
+open matrix finite_dimensional fintype polynomial finset
+
+namespace algebra
+
+variables (A : Type u) {B : Type v} (C : Type z) {ι : Type w}
+variables [comm_ring A] [comm_ring B] [algebra A B] [comm_ring C] [algebra A C]
+
+section matrix
+
+/-- Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define
+`trace_matrix A ι b` as the matrix whose `(i j)`-th element is the trace of `b i * b j`. -/
+@[simp] noncomputable
+def trace_matrix (b : ι → B) : matrix ι ι A
+| i j := trace_form A B (b i) (b j)
+
+lemma trace_matrix_def (b : ι → B) : trace_matrix A b = λ i j, trace_form A B (b i) (b j) := rfl
+
+variable {A}
+
+lemma trace_matrix_of_matrix_vec_mul [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+  trace_matrix A ((P.map (algebra_map A B)).vec_mul b) = Pᵀ ⬝ (trace_matrix A b) ⬝ P :=
+begin
+  ext α β,
+  rw [trace_matrix, vec_mul, dot_product, vec_mul, dot_product, mul_apply, bilin_form.sum_left,
+    fintype.sum_congr _ _ (λ (i : ι), @bilin_form.sum_right _ _ _ _ _ _ _ _
+    (b i * P.map (algebra_map A B) i α) (λ (y : ι), b y * P.map (algebra_map A B) y β)), sum_comm],
```

Comment:
Can you avoid the `@` if you use a `simp_rw` here? I haven't tried.

Reply:
It doesn't seem to work
```lean
simplify tactic failed to simplify
```